plasma_framework/test/src/exits/payment/PaymentChallengeIFEInputSpent.test.js

```diff
@@ -0,0 +1,357 @@
+const PaymentInFlightExitRouter = artifacts.require('PaymentInFlightExitRouterMock');
+const PaymentStartInFlightExit = artifacts.require('PaymentStartInFlightExit');
+const PaymentPiggybackInFlightExit = artifacts.require('PaymentPiggybackInFlightExit');
+const PaymentChallengeIFENotCanonical = artifacts.require('PaymentChallengeIFENotCanonical');
+const PaymentChallengeIFEInputSpent = artifacts.require('PaymentChallengeIFEInputSpent');
+const SpendingConditionMock = artifacts.require('SpendingConditionMock');
+const SpendingConditionRegistry = artifacts.require('SpendingConditionRegistry');
+const SpyPlasmaFramework = artifacts.require('SpyPlasmaFrameworkForExitGame');
+const StateTransitionVerifierAccept = artifacts.require('StateTransitionVerifierAccept');
+const ExitId = artifacts.require('ExitIdWrapper');
+
+const {
+    BN, constants, expectEvent, expectRevert, time,
+} = require('openzeppelin-test-helpers');
+const { expect } = require('chai');
+const {
+    TX_TYPE, OUTPUT_TYPE, EMPTY_BYTES, EMPTY_BYTES_32,
+} = require('../../../helpers/constants.js');
+const { buildUtxoPos } = require('../../../helpers/positions.js');
+const { createInputTransaction, createInFlightTx, getOutputId } = require('../../../helpers/ife.js');
+const { PaymentTransactionOutput, PaymentTransaction } = require('../../../helpers/transaction.js');
+const { spentOnGas, computeNormalOutputId } = require('../../../helpers/utils.js');
+
+contract('PaymentChallengeIFEInputSpent', ([_, alice, inputOwner, outputOwner, challenger]) => {
+    const IN_FLIGHT_EXIT_BOND = 31415926535; // wei
+    const PIGGYBACK_BOND = 31415926535;
+    const MIN_EXIT_PERIOD = 60 * 60 * 24 * 7; // 1 week
+    const DUMMY_INITIAL_IMMUNE_VAULTS_NUM = 0;
+    const INITIAL_IMMUNE_EXIT_GAME_NUM = 1;
+    const YOUNGEST_POSITION_BLOCK = 1000;
+    const INFLIGHT_EXIT_YOUNGEST_INPUT_POSITION = buildUtxoPos(YOUNGEST_POSITION_BLOCK, 0, 0);
+    const ETH = constants.ZERO_ADDRESS;
+    const BLOCK_NUMBER = 5000;
+    const MAX_INPUT_SIZE = 4;
+
+    before('deploy and link with controller lib', async () => {
+        const startInFlightExit = await PaymentStartInFlightExit.new();
+        const piggybackInFlightExit = await PaymentPiggybackInFlightExit.new();
+        const challengeInFlightExitNotCanonical = await PaymentChallengeIFENotCanonical.new();
+        const challengeIFEInputSpent = await PaymentChallengeIFEInputSpent.new();
+
+        await PaymentInFlightExitRouter.link('PaymentStartInFlightExit', startInFlightExit.address);
+        await PaymentInFlightExitRouter.link('PaymentPiggybackInFlightExit', piggybackInFlightExit.address);
+        await PaymentInFlightExitRouter.link('PaymentChallengeIFENotCanonical', challengeInFlightExitNotCanonical.address);
+        await PaymentInFlightExitRouter.link('PaymentChallengeIFEInputSpent', challengeIFEInputSpent.address);
+
+        this.exitIdHelper = await ExitId.new();
```

Comment:
Minor: I tends to put this separate with the others. We can have several `before('xxxx', async () => {...})` so we can express them differently. The reason I'd like to separate is that I found out current test code style would need to add links whenever new controller is introduced. As a result, this is the part of code that needs to alter, and it's easier to keep them the same so we can copy-paste easily :P

example code pointer: [here](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/test/src/exits/payment/PaymentPiggybackInFlightExitOnOutput.test.js#L41-L54)